constraints/hasher.air

```diff
@@ -10,49 +10,242 @@
 # 4. Merkle path verification.
 # 5. Merkle root update.
 #
-# STATUS: Not implemented
+# STATUS: Partially implemented (missing bus interactions)
 #
 # REFERENCES:
 # - Hasher chiplet design: https://0xmiden.github.io/miden-vm/design/chiplets/hasher.html
 ##########################################################################################
 
 mod hasher
 
+use utils::*;
+use rpo::enforce_rpo_round;
+
 ##########################################################################################
-# PERIODIC COLUMNS
+# HASHER CHIPLET TRANSITION CONSTRAINTS
+##########################################################################################
+
+# Enforces the constraints on the hash chiplet given its columns.
+ev hash_chiplet([s[3], h[12], i]) {
+    # Selector columns constraints
+    enf selector_columns([s]);
+
+    # Node index constraints
+    enf node_index([s, i]);
+
+    # Hasher state constraints
+    enf hasher_state([s, h, i]);
+}
+
+##########################################################################################
+# HELPER EVALUATORS
+##########################################################################################
+
+# Enforce selector columns constraints
+ev selector_columns([s[3]]) {
+
+    let f_out = get_f_out(s);
+    let f_out_next = get_f_out_next(s);
+    let f_abp = get_f_abp(s);
+    let f_mpa = get_f_mpa(s);
+    let f_mva = get_f_mva(s);
+    let f_mua = get_f_mua(s);
+
+    # Flag that is true when the performed operation is one of the operations represented by flags
+    # f_abp, f_mpa, f_mva or f_mua
+    let f_comp = f_abp + f_mpa + f_mva + f_mua;
+
+    # Enforce that selector columns are binary.
+    enf is_binary([selector]) for selector in s;
+
+    # Enforce that unless f_out = 1 or f_out' = 1, the values in columns s[1] and s[2] are copied 
+    # over to the next row.
+    enf is_unchanged([s[1]]) when !f_out & !f_out_next;
+    enf is_unchanged([s[2]]) when !f_out & !f_out_next;
+
+    # Enforce that if any of f_abp, f_mpa, f_mva, f_mua flags is set to 1, the next value of s[0] 
+    # is 0.
+    enf s[0]' * f_comp = 0;
```

Comment:
If I understand correctly, if we're in the last row of an operation which carries over to the next block, then the next row (which is the first of the next block), we must perform an output?

I think I see now though. We're indicating with `s[0]` that this is a block in which we will output the value when we reach the end?